### PR TITLE
Adding option for accumulating gradient

### LINKED ISSUE #437: Option to update policy with accumulated gradients
- [X] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [X] new feature request
- [X] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [X] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
Recently I've been running sub-policies in a wrapper. To keep maximum modularity in my code and make it easier to sub for different sub-policies, the ideal behavior is that the wrapper simply calls the learn() function of the sub-policies. However, they can't be updated simultaneously, since learn() does the gradient step by default. This might be slightly problematic in use cases such as mine, where sub-policies need to be updated simultaneously. (I can still get this behavior by wrapping yet another policy class on top of the Tianshou built-in policy classes, but this losses the modularity)

I propose a very simple additional feature: User can pass flags such as "accumulate_grad=True" as keyword arguments to learn() or update() method calls, and optimizer steps are not executed if this flag is checked to be true. The user now only need to make sure all update() calls are accumulating gradient, and do a final optimizer step in their own wrapper class.

### PR DESCRIPTION
**Pull request by Aceticia**
- [X] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [X] documentation modification
    + [X] new feature
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the code using `make commit-checks` (**required**)
- [ ] If applicable, I have mentioned the relevant/related issue(s)
- [ ] If applicable, I have listed every items in this Pull Request below

### CODE REVIEW

File: tianshou/policy/base.py [hunk 1]
```diff
@@ -185,6 +185,11 @@ def process_fn(
     def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, Any]:
         """Update policy with a given batch of data.
 
+        If you only want :meth:`learn` calls to accumulate gradients and not
+        actually update the weight, you can specify ``accumulate_grad=True``
+        in the argument. This doesn't block calls to function calls like
+        ``sync_weight``, please use with caution.
```
Comment (Trinkle23897) [inline tianshou/policy/base.py:191]:
```suggestion
        ``sync_weight``, please use with caution (see issue#437).
```

**Review by Trinkle23897 — COMMENTED** [review]
So is it possible to directly add some hooks in optimizer? Like the wrapper approach outside optim.zero_grad()and optim.step(), so that there's no need to add this kwargs everywhere...?
Or at least pass in a NULL optimizer which does nothing.

File: tianshou/policy/base.py (hunk 1 above)
Comment (Aceticia) [reply tianshou/policy/base.py:191]: Directly passing a dummy optimizer is an easy way to deal with this, but it will take away the choice of "whether to accumulate grads". Can you elaborate more on this: 

> Like the wrapper approach outside optim.zero_grad()and optim.step()

Comment (Trinkle23897) [reply tianshou/policy/base.py:191]: > it will take away the choice of "whether to accumulate grads"

I mean you can set up your choice by passing different `optim` into `policy` class



> Can you elaborate more on this:

For example, create a new optimizer that inherit existing pytorch optimizer (e.g., Adam):

```python
class MyAwesomeOptim(torch.optim):

  def __init__(self, accumulate_grad = True, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.opt = accumulate_grad

  def zero_grad(self):
    if self.opt:
        super().zero_grad()

  def step(self):
    if self.opt:
      super().step()
```

But I think the other approach is more concise and better.

Comment (Aceticia) [reply tianshou/policy/base.py:191]: I agree, the optimizer idea is more concise and looks way better. I'll close this pr then?

Comment (Trinkle23897) [reply tianshou/policy/base.py:191]: Sure!
